File: web/singularity-frontend/leaderboard.js
/**
 * leaderboard.js — fully client-side, no /api/* calls
 */

let walletPubkey = null;
let currentTab   = 'daily';

// Static mock traders (10 entries)
const BASE_TRADERS = [
    { addr: 'CrypK...1247', pnl: 2847.32,  roi: 24.3,  trades: 15,  winRate: 86.7, followers: 1247 },
    { addr: 'SolW...0892',  pnl: 1923.45,  roi: 19.7,  trades: 23,  winRate: 78.3, followers: 892  },
    { addr: 'DeFi...0634',  pnl: 1456.78,  roi: 15.2,  trades: 31,  winRate: 74.2, followers: 634  },
    { addr: 'TknH...0456',  pnl: 1234.56,  roi: 12.8,  trades: 18,  winRate: 72.2, followers: 456  },
    { addr: 'Meme...0321',  pnl:  987.65,  roi: 11.4,  trades: 27,  winRate: 70.4, followers: 321  },
    { addr: 'Arb7...0289',  pnl:  876.43,  roi:  9.8,  trades: 42,  winRate: 68.9, followers: 289  },
    { addr: 'Grid...0201',  pnl:  754.21,  roi:  8.3,  trades: 56,  winRate: 66.1, followers: 201  },
    { addr: 'Moon...0178',  pnl:  632.10,  roi:  7.1,  trades: 19,  winRate: 63.2, followers: 178  },
    { addr: 'Diam...0134',  pnl:  521.87,  roi:  5.9,  trades: 33,  winRate: 60.6, followers: 134  },
    { addr: 'Whal...0098',  pnl:  412.34,  roi:  4.7,  trades: 11,  winRate: 54.5, followers: 98   }
];

const PERIOD_MULTIPLIERS = { daily: 1, weekly: 7, monthly: 30, 'all-time': 365 };

const TOP_STRATEGIES = [
    { name: 'DCA Accumulation', users: 342, roi: 34.2 },
    { name: 'Momentum Scalp',   users: 218, roi: 28.7 },
    { name: 'Grid Trading',     users: 189, roi: 22.1 },
    { name: 'Arbitrage Bot',    users: 97,  roi: 18.4 },
    { name: 'Trend Following',  users: 156, roi: 15.9 }
];

const MARKET_LEADERS = [
    { token: 'SOL',  leader: 'CrypK...1247', pnl: 12400 },
    { token: 'JUP',  leader: 'DeFi...0634',  pnl: 8700  },
    { token: 'BONK', leader: 'Meme...0321',  pnl: 6200  },
    { token: 'RAY',  leader: 'Grid...0201',  pnl: 4800  },
    { token: 'S-IO', leader: 'SolW...0892',  pnl: 3100  }
];

document.addEventListener('DOMContentLoaded', () => {
    document.getElementById('wallet-btn').addEventListener('click', handleWalletClick);
    renderLeaderboard();
    renderTopStrategies();
    renderMarketLeaders();

    // Auto-connect
    setTimeout(async () => {
        if (window.solana?.isPhantom) {
            try {
                const r = await window.solana.connect({ onlyIfTrusted: true });
                walletPubkey = r.publicKey.toString();
                onConnect(walletPubkey);
            } catch {}
        }
    }, 600);
});

async function handleWalletClick() {
    if (walletPubkey) {
        await window.solana?.disconnect();
        walletPubkey = null;
        const btn = document.getElementById('wallet-btn');
        btn.textContent = 'Connect Wallet';
        btn.classList.remove('connected');
        document.getElementById('balance-display').classList.add('hidden');
        renderLeaderboard();
        return;
    }
    if (!window.solana?.isPhantom) { alert('Install Phantom Wallet'); return; }
    try {
        const r = await window.solana.connect();
        walletPubkey = r.publicKey.toString();
        onConnect(walletPubkey);
    } catch (e) { console.error(e); }
}

function onConnect(pub) {
    const btn = document.getElementById('wallet-btn');
    btn.textContent = `${pub.slice(0,4)}...${pub.slice(-4)}`;
    btn.classList.add('connected');
    document.getElementById('balance-display').classList.remove('hidden');
    window.loadWalletBalances(pub);
    renderLeaderboard();
}

function switchTab(tab) {
    currentTab = tab;
    document.querySelectorAll('.filter-tab').forEach(t => t.classList.remove('active'));
    event.target.classList.add('active');
    renderLeaderboard();
}

function renderLeaderboard() {
    const mult = PERIOD_MULTIPLIERS[currentTab] || 1;
    // Scale mock data by period multiplier with slight randomness
    let traders = BASE_TRADERS.map((t, i) => ({
        rank: i + 1,
        addr: t.addr,
        pnl:      +(t.pnl  * mult * (0.9 + Math.random() * 0.2)).toFixed(2),
        roi:      +(t.roi  * Math.sqrt(mult) * (0.9 + Math.random() * 0.2)).toFixed(1),
        trades:   Math.round(t.trades * mult * (0.8 + Math.random() * 0.4)),
        winRate:  +(t.winRate * (0.97 + Math.random() * 0.06)).toFixed(1),
        followers: t.followers
    }));

    // Insert user's wallet at random rank if connected
    if (walletPubkey) {
        let botStats = { total: 0, success: 0, pnl: 0 };
        try { botStats = JSON.parse(localStorage.getItem('bot-stats') || '{}'); } catch {}
        const userRank = Math.floor(Math.random() * 5) + 4; // rank 4-8
        const userEntry = {
            rank: userRank,
            addr: `${walletPubkey.slice(0,4)}...${walletPubkey.slice(-4)} (You)`,
            pnl:      +(botStats.pnl || 0).toFixed(2),
            roi:      +((botStats.pnl || 0) / 10).toFixed(1),
            trades:   botStats.total || 0,
            winRate:  botStats.total > 0 ? +((botStats.success / botStats.total) * 100).toFixed(1) : 0,
            followers: 0,
            isUser: true
        };
        traders.splice(userRank - 1, 0, userEntry);
        traders = traders.slice(0, 11).map((t, i) => ({ ...t, rank: i + 1 }));
    }

    const tbody = document.getElementById('leaderboard-body');
    tbody.innerHTML = traders.map(t => `
        <tr style="${t.isUser ? 'background:rgba(220,38,38,0.1);' : ''}">
            <td class="rank">#${t.rank}</td>
            <td style="color:${t.isUser ? '#ef4444' : '#fff'};">${t.addr}</td>
            <td class="${t.pnl >= 0 ? 'profit' : 'loss'}">$${t.pnl.toLocaleString()}</td>
            <td class="${t.roi >= 0 ? 'profit' : 'loss'}">${t.roi >= 0 ? '+' : ''}${t.roi}%</td>
            <td>${t.trades}</td>
            <td>${t.winRate}%</td>
            <td>${t.followers}</td>
            <td><button onclick="followTrader('${t.addr}')" style="padding:.4rem .9rem;background:#dc2626;color:#fff;border:none;border-radius:4px;cursor:pointer;font-size:.85rem;">Follow</button></td>
        </tr>`).join('');
}

function renderTopStrategies() {
    const html = TOP_STRATEGIES.map(s => `
        <div style="display:flex;justify-content:space-between;padding:.8rem 0;border-bottom:1px solid #333;">
            <div>
                <div style="color:#fff;font-weight:bold;">${s.name}</div>
                <div style="color:#666;font-size:.9rem;">${s.users} users</div>
            </div>
            <div style="color:#ef4444;font-weight:bold;">+${s.roi}%</div>
        </div>`).join('');
    const el = document.getElementById('top-strategies');
    if (el) el.innerHTML = html;
}

function renderMarketLeaders() {
    const html = MARKET_LEADERS.map(l => `
        <div style="display:flex;justify-content:space-between;padding:.8rem 0;border-bottom:1px solid #333;">
            <div>
                <div style="color:#dc2626;font-weight:bold;">${l.token}</div>
                <div style="color:#ccc;font-size:.9rem;">${l.leader}</div>
            </div>
            <div style="color:#ef4444;font-weight:bold;">+$${(l.pnl/1000).toFixed(1)}K</div>
        </div>`).join('');
    const el = document.getElementById('market-leaders');
    if (el) el.innerHTML = html;
}

function followTrader(addr) {
    alert(`Now following ${addr}!`);
}

File: api/leaderboard.py
from fastapi import APIRouter
from typing import Dict, List
from datetime import datetime, timedelta
import random

router = APIRouter()

# Mock trading data - in production this would come from a database
trading_data = {
    'daily': [
        {'rank': 1, 'name': 'CryptoKing', 'pnl': 2847.32, 'roi': 24.3, 'trades': 15, 'winRate': 86.7, 'followers': 1247},
        {'rank': 2, 'name': 'SolanaWhale', 'pnl': 1923.45, 'roi': 19.7, 'trades': 23, 'winRate': 78.3, 'followers': 892},
        {'rank': 3, 'name': 'DeFiMaster', 'pnl': 1456.78, 'roi': 15.2, 'trades': 31, 'winRate': 74.2, 'followers': 634},
        {'rank': 4, 'name': 'TokenHunter', 'pnl': 1234.56, 'roi': 12.8, 'trades': 18, 'winRate': 72.2, 'followers': 456},
        {'rank': 5, 'name': 'MemeTrader', 'pnl': 987.65, 'roi': 11.4, 'trades': 27, 'winRate': 70.4, 'followers': 321}
    ],
    'weekly': [
        {'rank': 1, 'name': 'CryptoKing', 'pnl': 15847.32, 'roi': 124.3, 'trades': 89, 'winRate': 84.3, 'followers': 1247},
        {'rank': 2, 'name': 'DeFiMaster', 'pnl': 12456.78, 'roi': 98.7, 'trades': 156, 'winRate': 76.9, 'followers': 634},
        {'rank': 3, 'name': 'SolanaWhale', 'pnl': 9823.45, 'roi': 87.2, 'trades': 134, 'winRate': 73.1, 'followers': 892}
    ],
    'monthly': [
        {'rank': 1, 'name': 'CryptoKing', 'pnl': 67847.32, 'roi': 456.7, 'trades': 423, 'winRate': 82.7, 'followers': 1247},
        {'rank': 2, 'name': 'DeFiMaster', 'pnl': 54321.09, 'roi': 387.4, 'trades': 567, 'winRate': 78.2, 'followers': 634}
    ],
    'all-time': [
        {'rank': 1, 'name': 'CryptoKing', 'pnl': 234567.89, 'roi': 1234.5, 'trades': 2341, 'winRate': 81.4, 'followers': 1247},
        {'rank': 2, 'name': 'DeFiMaster', 'pnl': 187654.32, 'roi': 987.6, 'trades': 1876, 'winRate': 77.8, 'followers': 634}
    ]
}

strategies_data = [
    {'name': 'DCA Bot', 'roi': 24.3, 'users': 156},
    {'name': 'Momentum Trading', 'roi': 19.7, 'users': 89},
    {'name': 'Arbitrage', 'roi': 15.2, 'users': 234},
    {'name': 'Grid Trading', 'roi': 12.8, 'users': 67}
]

market_leaders = [
    {'token': 'SOL', 'leader': 'CryptoKing', 'pnl': 2800},
    {'token': 'USDC', 'leader': 'DeFiMaster', 'pnl': 1900},
    {'token': 'RAY', 'leader': 'SolanaWhale', 'pnl': 1500},
    {'token': 'ORCA', 'leader': 'TokenHunter', 'pnl': 1200}
]

@router.get("/api/leaderboard/{period}")
async def get_leaderboard(period: str):
    """Get leaderboard for specified time period"""
    if period not in trading_data:
        period = 'daily'
    
    # Add some randomness to simulate live data
    data = trading_data[period].copy()
    for trader in data:
        trader['pnl'] += random.uniform(-50, 50)
        trader['roi'] += random.uniform(-1, 1)
    
    return {
        "period": period,
        "leaderboard": data,
        "updated_at": datetime.now().isoformat()
    }

@router.get("/api/leaderboard/strategies")
async def get_top_strategies():
    """Get top performing trading strategies"""
    # Add randomness
    data = strategies_data.copy()
    for strategy in data:
        strategy['roi'] += random.uniform(-2, 2)
        strategy['users'] += random.randint(-5, 10)
    
    return {
        "strategies": data,
        "updated_at": datetime.now().isoformat()
    }

@router.get("/api/leaderboard/market-leaders")
async def get_market_leaders():
    """Get market leaders by token"""
    # Add randomness
    data = market_leaders.copy()
    for leader in data:
        leader['pnl'] += random.uniform(-100, 200)
    
    return {
        "market_leaders": data,
        "updated_at": datetime.now().isoformat()
    }

@router.post("/api/leaderboard/follow/{trader_name}")
async def follow_trader_leaderboard(trader_name: str, wallet_data: dict):
    """Follow a trader from leaderboard"""
    # This would integrate with the social trading system
    return {
        "success": True,
        "message": f"Now following {trader_name}! Copy trading activated.",
        "trader": trader_name,
        "wallet": wallet_data.get("wallet")
    }